spark/src/test/java/org/apache/iceberg/spark/data/TestHelpers.java

```diff
@@ -267,11 +267,23 @@ private static void assertEqualsUnsafe(Type type, Object expected, Object actual
     }
 
     switch (type.typeId()) {
-      case BOOLEAN:
       case INTEGER:
+        Assert.assertEquals("Values didn't match", ((Number) expected).intValue(),
+                ((Number) actual).intValue());
```

Comment:
Why was it necessary to change this method? Shouldn't these values be the same type already?

Reply:
Without this change, the comparison fails with ClassCastException.

Reply:
Sounds like you're comparing objects that are two different types. Can you find out why the types don't match?

Reply:
Yeah. The records in this test are created with data types int and float. However, when we read it, we use the Iceberg schema which has these columns as long and double.

Reply:
Okay, so the problem is that the expected records used to match have not been promoted?

If that's the case, then I think we should handle type promotion for the expected values in the assertions. Let's pass the expected schema, which will have `long` or `double`. Then the behavior for `int` and `float` don't need to change because there is no promotion to handle. In the cases for `long` and `double`, the expected value can be checked using `expected instanceof Integer` and promoted if needed. Otherwise, the original `assertEquals` can be used.

I think that will more narrowly validate the cases here. What we want to avoid is allowing `1` and `1.0D` to match if the type is `int` because there is no promotion to `int` and a `double` should never be promoted to `long` if the type were `long`.